src/analysis/retail/evoker/preservation/modules/talents/TimeOfNeed.tsx

```diff
@@ -0,0 +1,241 @@
+        //Check that ToN actually did any healing
+        if (spawn.verdantEmbrace !== null || spawn.livingFlames.length > 0) {
+          //If there is no damage events, then Time of Need didn't save the player from anything. Otherwise, parse them
+          if (eventList.events.length !== 0) {
+            //Iterate all the damage taken events during the time window
+            eventList.events.forEach((hit) => {
+              hit = hit as DamageEvent;
+              //Check if the hit landed on the same player that the Time of Need targeted
+              if (hit.targetID === spawn.target) {
+                //If the hitpoints after the hit are equal or less than 0, the player died anyway
+                if (hit.hitPoints !== undefined && hit.hitPoints === 0) {
+                  this.spawns[index].result = Result.Dead;
+                } else {
+                  //Initialize the healing from Time of Need with the VE
+                  let tonHealingAtThisPoint = spawn.verdantEmbrace?.amount || 0;
+                  //If there is no living flames casted, or this hit was before the first living flame landed
+                  if (
+                    spawn.livingFlames.length === 0 ||
+                    hit.timestamp < spawn.livingFlames[0].timestamp
+                  ) {
+                    //Check if the player would've died from this hit without the Verdant Embrace healing
+                    if (
+                      hit.hitPoints &&
+                      hit.hitPoints > 0 &&
+                      hit.hitPoints - tonHealingAtThisPoint <= 0
+                    ) {
+                      this.spawns[index].result = Result.Proc; //Person was saved by the VE
+                    }
+                  } else {
+                    //Add up all the healing done by the living flames that happened before this hit
+                    spawn.livingFlames.forEach((livingFlame) => {
+                      if (livingFlame.timestamp < hit.timestamp) {
+                        tonHealingAtThisPoint += livingFlame.amount;
+                      }
+                    });
+                    //If the hit points after this hit minus the total ToN healing done to this point is equal or less than 0, then he was saved
+                    if (
+                      hit.hitPoints &&
+                      hit.hitPoints > 0 &&
+                      hit.hitPoints - tonHealingAtThisPoint <= 0
+                    ) {
+                      //Mark as long save
+                      this.spawns[index].result = Result.Long;
+                    }
+                  }
+                }
+              }
+            });
+          }
+        } else {
+          //If ToN did no healing, it bugged out
+          this.spawns[index].result = Result.Bug;
+        }
+      }
+    });
+  }
+
+  async load() {
+    //Run every ToN spawn through parseDamageTaken()
+    for (const spawn of this.spawns) {
```

Comment:
how many ToN spawns would you expect in a typical M+ run? potentially worried about hitting burst limits if there are hundreds of possible spawns in a long key.

looking at [this log](https://www.warcraftlogs.com/reports/KGaV46B9LrFHhk8W#fight=1&type=healing&source=18&translate=true) i see an avg of like 1 every 2m, which is probably fine?

Reply:
The talent itself has a 1min cd so theoretical maximun is as many minutes as the key had, realistically i've seen between 9 and 16 spawns per key

Reply:
```tsx
    for (const spawn of this.spawns) {
      const target = this.combatants?.players[spawn.target];
      const filter = target && target?.name ? `target.name = "${target.name}"` : '';
      const damageTakenEvents = await fetchWcl<WCLEventsResponse>(
        `report/events/damage-taken/${this.owner.report.code}`,
        {
          start: spawn.summon.timestamp,
          end: spawn.summon.timestamp + 8000,
          undefined,
          filter,
        },
      );
```
You could potentially add this filter to reduce the amount of events you query

Reply:
(filter is a good addition)

Reply:
the undefined param is optional `actorId` iirc, but suppose you could just do `filter: filter` to avoid the undefined

Reply:
since this is an object literal, `filter` is shorthand for `filter: filter`. I don't believe you need to explicitly handle the actor id param, so omitting the undefined should be fine. If other comments are addressed I can handle that and get it merged

Reply:
All the other comments got addressed (and also the filter with the 'undefined') in my last 2 commits